File: quiz/deno/src/statemaschine/states/login-state.ts
import { Input, Secret } from '@cliffy/prompt/mod.ts';
import { User } from '@src/user.ts';
import { LoggedInState } from '@states/logged-in/logged-in-state.ts';
import { IState } from '@states/state.ts';
import { UI } from '@src/utils.ts';

export class LoginState implements IState {
	async run(): Promise<IState> {
		const username = await Input.prompt('Username: ');
		const password = await Secret.prompt('Password: ');

		try {
			const user = await User.checkLogin(username, password);
			return new LoggedInState(user);
		} catch {
			UI.prompt(
				'%cInvalid username or password',
				'color: red',
			);
			await UI.pause(2);
			return new LoginState();
		}
	}
}

File: quiz/go/statemaschine/states/login-state.go
package states

import (
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/user"
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/utils"
	"github.com/manifoldco/promptui"
)

type LoginState struct{}

func (state *LoginState) Run() IState {
	username := state.promptUsername()
	password := state.promptPassword()

	u, err := user.CheckLogin(username, password)
	if err != nil {
		panic(err)
	}

	if u == nil {
		utils.Prompt("Invalid username or password")
		utils.Pause(2)
		return &LoginState{}
	} else {
		return &LoggedInState{user: u}
	}
}

func (state *LoginState) promptUsername() string {
	prompt := promptui.Prompt{
		Label: "Username",
	}

	result, _ := prompt.Run()
	return result
}

func (state *LoginState) promptPassword() string {
	prompt := promptui.Prompt{
		Label: "Password",
		Mask:  '*',
	}

	result, _ := prompt.Run()
	return result
}
